index.js

```diff
@@ -60,40 +79,73 @@ class Cache {
   }
 }
 
-let _currentSecond
-
-function currentSecond () {
-  if (_currentSecond !== undefined) {
-    return _currentSecond
-  }
-  _currentSecond = Math.floor(Date.now() / 1000)
-  setTimeout(_clearSecond, 1000).unref()
-  return _currentSecond
-}
-
-function _clearSecond () {
-  _currentSecond = undefined
-}
-
 class Wrapper {
-  constructor (func, key, serialize, cacheSize, ttl, onHit) {
-    this.ids = new LRUCache(cacheSize)
-    this.error = null
-    this.started = false
+  // TODO signature
+  constructor (func, key, serialize, references, /*size, */storage, ttl, onDedupe, onHit, onMiss) {
+    // TODO do we want to limit dedupe size?
+    this.dedupes = new Map()
     this.func = func
     this.key = key
     this.serialize = serialize
+    this.references = references
+
+    this.storage = storage
     this.ttl = ttl
-    this.onHit = onHit
+    this.onDedupe = onDedupe // TODO bind data
+    this.onHit = onHit // TODO bind data
+    this.onMiss = onMiss // TODO bind data
+  }
+
+  async wrapFunction (args, key) {
+    return async () => {
+      const data = await this.storage.get(key)
+      if (data) {
+        this.onHit()
+        return data
+      }
+
+      this.onMiss()
+
+      // TODO check this block for leaks and issues
+      // it should be safe because it's wrapped by query.promise
+      const result = await this.func(args, key)
+
+      if (this.ttl < 1) {
+        return result
+      }
+
+      if (!this.references) {
+        await this.storage.set(key, result, this.ttl)
+        return result
+      }
+
+      const references = this.references(args, key, result)
+      await this.storage.set(key, result, this.ttl, references)
+
+      return result
+    }
   }
 
   buildPromise (query, args, key) {
-    query.promise = this.func(args, key)
+    // wrap the original func to sync storage
+    // TODO move to a function
+    query.promise = this.wrapFunction(args, key)
```

Comment:
I think you can simplify the `wrapFunction` and remove the returned function. Note that the current code cannot work as-is.